### PR TITLE
Report: Add timing percentiles

### PR DESCRIPTION
**Pull request by TBonnin**
Notification report endpoins now returns the timing percentiles rather than a map of all timings 10secs slots

screen shot 2018-10-15 at 11 11 01

### CODE REVIEW

File: commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala [hunk 1]
```diff
@@ -4,13 +4,54 @@ import java.time.temporal.{ChronoUnit, Temporal, TemporalUnit}
...
   providerCounts: ProviderCount,
   timing: Map[LocalDateTime, Int]
-)
+) {
+
+  def timingPercentiles: Option[TimingPercentiles] = {
+    val allTimings: Seq[LocalDateTime] = timing
+      .toSeq
+      .flatMap { case (t, count) =>
+        Seq.fill(count)(t)
```
Comment (TBonnin) [inline commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala]: This is quite heavy memory-wise. @alexduf Can you think of a better way to calculate percentiles?

File: commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala [hunk 2]
```diff
@@ -79,6 +120,12 @@ object EventAggregation {
...
     (dateTimeMap: Map[LocalDateTime, Int]) => Json.toJsObject(dateTimeMap.map { case (k, v) => (k.toString, v) })
   )
-  implicit val jf = Json.format[EventAggregation]
+
+  implicit val jreads = Json.reads[EventAggregation]
+  implicit val jwrites: Writes[EventAggregation] =
+    Json
+      .writes[EventAggregation]
+      .addField("timingPercentiles", _.timingPercentiles)
+      .removeField("timing")
```
Comment (TBonnin) [inline commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala]: I made the decision to hide the timing map. It is super long and doesn't provide much information as is.

File: commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala (hunk 1 above)
Comment (alexduf) [reply commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala]: Yes, I think we have count of the total notification received.
If we take the timings and sort it, then iterate through it, we can stop when we have the total number of notification so far > than the percentile you're calculating. Something like that (definitely not working scala, just to get the logic)

```scala
val mySortedTimings: Seq[(LocalDateTime, Int)] = timing.toSeq.sort
val `80thPercentileCount` = total * 0.8
val (`80thPercentileTime`, _) = mySortedTimings.foldLeft((notificationSentTime, 0)) {
  case ((_, totalSoFar), (time, count)) if totalSoFar < `80thPercentileCount` => (time, count + totalSoFar)
  case ((time, totalSoFar), _) if totalSoFar >= `80thPercentileCount` => (time, totalSoFar)
}
```

File: commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala (hunk 2 above)
Comment (alexduf) [reply commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala]: No I agree! we just need to double check if that has an impact on the event consumer that stores the data in dynamo

**Comment by alexduf** [comment]
Also, should the percentile be relative to the sent time?

File: commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala (hunk 2 above)
Comment (alexduf) [reply commoneventconsumer/src/main/scala/com/gu/notifications/events/model/EventAggregation.scala]: I double-checked, it's okay to hide it like that :+1:

**Comment by TBonnin** [comment]
@alexduf I have refactored the percentiles calculations. See https://github.com/guardian/mobile-n10n/pull/216/files#diff-9d2d9b7cfea3578513d3888e49c5feb6R24
It requires to iterate through the map more than once but it doesn't create a gigantic list of all values. (tl;dr: less efficient computatin-wise but more efficient memory-wise)

As discussed on the chat, relative duration could be good but will require some more refactoring to store the notification sending time. Let's do that in another PR if we realise we really need it

**Pull request merged**